Register/login frontend

courtmcgee2197 opened the pull request:
Fixed urls for the user.
Login/Register Page are setup so we can test users
I made slight changes to Post to demonstrate the use of connecting users with backend and frontend. The code is still there just commented it out for testing purposes.
JWT Tokens allow for users to be logged in for about a day before needing to re login
The home page is essentially just a test to connect the user to their posts
The update page allows for updating username add more fields to be updated if necessary
Register page will add users to database
Login will handle authentication
There are many ways to access the current user so based on the app you are trying to implement will ultimately be based on the programmer how you implement

thecatfour commented:
Regarding the frontend directory structure, this pull request isn't following our agreed convention. What it should look like is something like:
```
src/
    app/
         Home/
             page.js
         Login/
         Register/
         Update/
    components/
         DirectoryTitleForComponents/
             componentFiles.js
    lib/
         constants or other functions.js
```
Basically, the components should be in a different directory than the app/ directory. We don't have a lib/ directory right now, so you can probably just skip it.

I would like the changes you made to backend/Posts to have comments explaining why they are commented out rather than removed. The pull request states that these changes are for testing purposes, but this should also be explained in the code. For example, the fields for the post was changed.

![image](https://github.com/user-attachments/assets/cf0f13d7-f0d6-4549-820f-731099545f6e)

I remember you saying that this change was for testing if the authentication needed for posting was working. If this change was just for your own testing, then you should undo the testing changes before we merge the pull request. 

**Must Haves**

- Move your components into the appropriate directories to conform with our design
- Undo the changes made to backend/Posts or explain any improvements in the code

**Nice To Have**

- Comment mentioning that ACCESS_TOKEN and REFRESH_TOKEN are keys for localStorage

**Nitpicks**

- The difference in whitespace in Home/page.js and the component files RegisterForms.js and LoginForms.js looks unusual

barrioskevin commented:
To add on to what leslie was saying about our projects structure you should add in a folder called `accounts/` or something along those lines where inside you include your update, login, register directories. You should also consider lowercase names because I fear the way you have it setup we will need to include caps in the url or at the very least seem them capital in the address bar. 

I think the home folder is unnecessary. Instead you should edit the `page.js` file thats in `src/app/` as this file is our true homepage. In this file you can make so it will redirect to register or login and then you can have just a blank page for authenticated users where we can see access to the `ACCESS_TOKEN` and `REFRESH_TOKEN`. (maybe just make an authenticated GET call to your UserDetailView in the backend where we send back data about the user like email and password)

We should probably leave messing with the posts api OUT of his pull request. You already added some great additions with making authenticated users and I believe that is enough for this pull request. The posts api is eventually gonna be updated as the rest of the pieces for our app come together so there's no need to make any changes at the moment. This includes changing the url for it as well.

Your `api.js` file looks like a really cool helper function along with your protected route component. I wish you explained this more in your initial post. I want to hear more about you plan on us using in throughout our app. You are also still not following the backend url convention where we add in `api/` the 2 apps already in our project use this notation. If you choose to change it then let the two people who created these apps, posts and calendar, know about it so they can update their url to not include `api/`. I still think it's better if we do include it. It's already in use with the two apps in our backend currently and it provides a clear distinction. 

in MY opinion for this PR to be merged
Must Have:
- Fix the structure for your folders so it follows what we agreed on.
- Leave any changes to post api and your create post demo out of your current PR.
- Provide a explanation or visual of your `api.js` and `ProtectedRoute.js` files
- Don't `CORS_ALLOW_ALL_ORIGINS = True` it is not good practice and it's not needed. We previous explained what you need to do if you can't access the backend. (add in your link). We want calls to the backend to only be coming from our app.   

Nice to have:
- Make your homepage additions to the actual projects home page `src/app/page.js`
- Add in more views or expand current functionality so we can make reads. (this can be done in the future as we add in more)
- Maybe instead of adding all that logic in your register form. Just make the backend only allow accounts to be created with a unlv email. It makes more sense to keep that logic there and in the frontend we can just handle the error.  

Nit-Picks:
- You add in a lot of new dependencies and also add in a lot of new logic in the front end with `api.js` and localstorage. I know that we don't have a set template right now but it would be nice for you to let us know why you made such changes and how we are expected to use them.

Overall really nice work I like how you handled authorized access with your protected routes component. Hopefully with a few adjustments we can get this merged quickly!

Commit:
Adjusted PR

courtmcgee2197 commented:
First I have fixed my directory structure to match our agreed convention.
Undid changes to Post code.
Fixed urls  to include api in the front.
Removed cors allow all origins.

This is how it will be used going forward 
This is the code for logging in the route being passed in api/users/token which is and endpoint that takes a username and password and will return two tokens the access and refresh and set it in local storage(think of it as a session or cookie).
<img width="670" alt="login" src="https://github.com/user-attachments/assets/c3c82f1c-0398-434e-b5e1-236b0813871b">
Now that we have our tokens in local storage we can make requests while we are logged in. This is the helper code that way we dont have to rewrite axios create everytime. Basically the code will grab the token currently in local storage and attach it to the request for backend to verify.
<img width="512" alt="axios request" src="https://github.com/user-attachments/assets/54e35f20-b581-43f8-b904-1eafcdbea13f">
Now in order for this to work there must be pages that cannot be accessed unless logged in. This is what the protected route is for. First the code checks if there is a token in local storage. Only time there wouldnt be is if you attempt to access a page without having logged in. Now typically we dont want users logged indefinitely and we also don't want the token available for too long just in case it is somehow leaked. This is the use for refresh token. First decode the jwt token check its expiration and if expired use the refreshtoken() function to get a new token. Currently I have it set to 30 minutes for access token and 24 hours for the refresh token. At any point if you are on protected page and both tokens have expired it will set isAuthorized to false and redirect to login 
<img width="639" alt="protectedRoute" src="https://github.com/user-attachments/assets/1e72452a-ea3f-4cbb-b1ad-861f5058e258">
Also all page.js that need to accessed by user should be done this way
<img width="340" alt="protectedpage" src="https://github.com/user-attachments/assets/d627c82a-7cbf-401e-aa43-90658e36ab66">

Now to demonstrate how to use it this code is not in the pull request by the way. Using the api function and create post endpoint we send the title and message and on the backend it will set the author to current user see here 
<img width="419" alt="author" src="https://github.com/user-attachments/assets/f9e85e0a-5a05-4b90-993b-82e8e9f60f0d">

<img width="503" alt="create a post" src="https://github.com/user-attachments/assets/0d1a8b33-e19f-4c6c-aa4b-f6d5ab267b9c">
<img width="958" alt="Demonstration" src="https://github.com/user-attachments/assets/b430bcef-bb79-476e-a49a-e6857ae78efc">
<img width="643" alt="createdemo" src="https://github.com/user-attachments/assets/4cac7c0f-9922-4edc-9767-b4be9c9bc070">

Now lets see how we can get all the posts associated with a user 
Here is front end code
<img width="568" alt="getposts associated with user" src="https://github.com/user-attachments/assets/8552ce1a-d0a6-43c3-b1c6-6465fe4685c7">
Here is what needs to be in backend
<img width="404" alt="backend user post list" src="https://github.com/user-attachments/assets/f3740546-51c0-4ea3-bf99-b182be3f4874">
The list will be an array so to display code has to look something like this 
<img width="532" alt="display html" src="https://github.com/user-attachments/assets/9d8f4021-1fdc-4948-bae6-0e9dbb5d876b">
Here is resulting page
<img width="542" alt="post list demo" src="https://github.com/user-attachments/assets/51b7e484-91ab-4034-b4d1-8b6f214e5db0">

Ok I do believe I resolved the issues and provided documentation on as how this will be used going forward

thecatfour commented:
Everything major seems to be fixed. I only have a nitpick with how our urlpatterns are becoming inconsistent. calendar_app and users follow the following naming pattern:

![image](https://github.com/user-attachments/assets/ea9fe195-6bce-4e1c-911d-0c6895599cf0)

![image](https://github.com/user-attachments/assets/147e4587-0724-48dd-a2ce-cf49d07f6ef5)

However, Posts adds the 'api/' part in rebel_hubs/urls.py:

![image](https://github.com/user-attachments/assets/cf260aa9-81c4-4c1d-b6e9-1b0b1e16a242)

I think we should agree on which formatting to use.

barrioskevin commented on backend/Posts/serializers.py:
```diff
@@ -4,7 +4,7 @@
 class PostSerializer(serializers.ModelSerializer):
     class Meta:
-        model = Post
+        #model = Post
```
this looks like an issue, is it possible to just leave changing this file out of the pr?

barrioskevin commented on backend/Posts/urls.py:
```diff
@@ -4,7 +4,7 @@
 
     # The PostList class used for getting all the post (GET) 
-    path('posts/', PostList.as_view(), name='post-list'),  
+    path('posts/', PostList.as_view(), name='post-list'),
```
looks identical

barrioskevin replied:
whitespace ig

barrioskevin commented on backend/Posts/views.py:
```diff
@@ -9,7 +9,7 @@
 # Able to create and view the post, should only handles POST requests to create a post
 class CreatePost(generics.CreateAPIView):
-    queryset = Post.objects.all()
+    #queryset = Post.objects.all()
```
again this looks like an issue can we leave any changes to these files out of your pr?

barrioskevin commented on backend/rebel_hubs/urls.py:
```diff
@@ -23,4 +23,5 @@
     path('api/', include('Posts.urls')),
     # path('api/', include('Comments.urls')), will be added in the next PR
     path('', include('calendar_app.urls')),
+    path('', include('users.urls')),
```
ideally we would add the api/ right here in order for your urls in users.urls to look a lot cleaner, it's just preference but if you look at how Posts does it it seems better

barrioskevin replied:
it would look something like `path('api/', include('users.urls')),` and you can remove your api/ prefix in your users.url but again thats totally up to you im just glad we are following a convention

courtmcgee2197 replied:
Im not sure that code even works this was the result 
<img width="551" alt="api" src="https://github.com/user-attachments/assets/e14d1180-f8aa-45f0-b242-c8b2141215bb">
<img width="442" alt="urls" src="https://github.com/user-attachments/assets/d93b41e3-6f56-486e-8847-ffa9f3af243f">

barrioskevin replied:
i went to test it out for myself and it works.
![image](https://github.com/user-attachments/assets/70e9b7be-78a4-4e1f-b364-032dc6cacb9c)
![image](https://github.com/user-attachments/assets/9049edc9-6862-4a16-b780-0bf6ea3429b6)
you would just need to make changes to the two url files. but again it's up to you we can make a pr in the future that strictly deals with cleaning up paths

barrioskevin commented on backend/users/tests.py:
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
+
+# Create your tests here.
```
at least one simple test should be required for when we eventually add in CI. but maybe right not it's not needed theres really only basic functionality.

Commit:
requirements.txt and time for tokens

Commit:
Altered file paths

barrioskevin commented on frontend/src/components/Accounts/RegisterForms.js:
```diff
@@ -0,0 +1,72 @@
+        try {
+            console.log(route)
+            const request = await api.post(route, { email,username, password})
+            router.push('/login')
```
`router.push('/users/login')` to fit your new directory structure

Commit:
Dependencies

barrioskevin commented on frontend/src/app/users/update/page.js:
```diff
@@ -0,0 +1,40 @@
+    const handleUpdate = async () => {
+
+        try {
+            const response = api.patch(`api/users/currentUser/`, {username:newUsername})
```
did you need to add an await here too? i didn't check that

thecatfour committed:
Update serializers.py

thecatfour committed:
Update views.py

thecatfour committed:
Update urls.py

thecatfour committed:
Update urls.py

thecatfour committed:
Update page.js

thecatfour committed:
Update ProtectedRoutes.js

thecatfour committed:
Update ProtectedRoutes.js

thecatfour committed:
Update RegisterForms.js

thecatfour committed:
Update urls.py
